### ISSUE TITLE
Themetype question

### ISSUE THREAD

**Issue by srd-software** [VuioVuio/DelphiUCL#3]
I'm having a great time playing around with this stuff. Question on TUThemeType. Would it be possible to add a user type? 

So from this:

```TUThemeType = (ttAuto, ttLight, ttDark);```

To this:

```TUThemeType = (ttAuto, ttLight, ttDark ,ttUser);```

So we'd have:
- ttAuto: System defined AccentColor
- ttLight: Predefined "light" AccentColor
- ttDark: Predefined "dark" AccentColor
- ttUser: User defined AccentColor (any color)

I've worked around things for now so I can get and set a user color in my test app:

![image](https://user-images.githubusercontent.com/51385541/77174944-72f4da80-6a87-11ea-9d5f-c0d161b14ff8.png)

**Comment by ghost** [comment]
ThemeManager provides the AccentColor property to get the current accent color. And to set a custom accent color, set the AccentColorType property (set to 0 to use system color). So I think it unnecessary to add a ttUser to TUThemeType.

**Comment by srd-software** [comment]
OK, I'll try that and see what I can do.

Thanks!

**Comment by srd-software** [comment]
OK, I've tried all sorts of things and I must be doing it wrong!

I would expect this to work and give me the system accent color:
```
  ThemeManager.ThemeType := ttAuto;
  ThemeManager.AccentColorType := 0;//clTeal;
  ThemeManager.UpdateTheme;
  UCaptionBar1.UpdateTheme(false);
```
But it just gives me the initial dark color of 2b2b2b. I was expecting that UCaptionBar1.UpdateTheme would update the colors but it never seems to do anything. The following will work to give me the system color:

```
  ThemeManager.ThemeType := ttAuto;
  ThemeManager.AccentColorType := 0;//clTeal;
  ThemeManager.UpdateTheme;
  UCaptionBar1.CustomBackColor.Enabled := true;
  UCaptionBar1.CustomBackColor.SetColor(0,0,ThemeManager.AccentColor);
```
That's fine, but then if I do this:
```
  ThemeManager.ThemeType := ttDark;
  //ThemeManager.AccentColorType := 0;//clTeal;
  ThemeManager.UpdateTheme;
  UCaptionBar1.CustomBackColor.Enabled := true;
  UCaptionBar1.CustomBackColor.SetColor(0,0,ThemeManager.AccentColor);
```
I would expect AccentColorType to be ignored and get 2b2b2b but even if I comment out the AccentColorType line like above, I still get the system accent color.

Obviously I'm not understanding what is intended so could you give me some tips? Basically what I want is the least amount of code to be able to set the caption color to:

- System (whatever the AccentColor is in the registry)
- Light (pre defined "light" theme)
- Dark (pre defined "dark" theme which appears to be 2b2b2b)
- User (any user defined color)

I've got this all working in several test apps but I'm thinking I'm going about it all the wrong way and doing a lot more work in the code than is necessary.

**Comment by srd-software** [comment]
OK, more testing. This seems to work pretty well:
```
  case ThemeManager.ThemeType of
    ttAuto: ThemeManager.AccentColorType := clTeal; //or 0 for system color
    ttLight: ThemeManager.AccentColorType := $00F2F2F2;
    ttDark: ThemeManager.AccentColorType := $002b2b2b;
  end;

  ThemeManager.UpdateTheme;

  //Set CAPTIONBAR_BACK instead of the caption control so UpdateTheme (below) works:
  CAPTIONBAR_BACK.SetColor(clGray,$00F2F2F2,ThemeManager.AccentColor);

  //Set colors for use when app deactivated:
  UCaptionBar1.CustomBackColor.LightColor := clSilver; //non-active color for ttLight
  UCaptionBar1.CustomBackColor.DarkColor := CAPTIONBAR_BACK.Color; //non-active color for everything else

  UCaptionBar1.UpdateTheme(false); //actually works if we set CAPTIONBAR_BACK 
```
When I need to have the colors switch between app active and inactive, all it takes now is:
```
UCaptionBar1.CustomBackColor.Enabled := not Active;
```
Does this seem to fit with the design intentions/logic of your code?

**Comment by ghost** [comment]
Yes, correct. All you need is to set the Enabled property of the CustomABC property to switch between the global theme set (CAPTIONBAR_BACK) and custom theme set (CustomBackColor property). You can use it as two different color sets for the active and inactive states of the form.

**Comment by ghost** [comment]
You should learn more about TUThemeColorSet in UCL.Colors unit. Basically, it is a collection of colors based on the current theme (auto, light, dark).
Each control may have a corresponding number of themeset. Here I will choose TUCaptionBar as an example.

The TUCaptionBar control has two theme sets: the CAPTIONBAR_BACK (global themeset) and the CustomBackColor (custom themeset) property.
Changing the color in CAPTIONBAR_BACK will affect all CaptionBar controls, but changing the CustomBackColor property will only apply to the current control.
CustomBackColor colors are copied from CAPTIONBAR_BACK when the object is created. And when drawing controls, it will choose between CAPTIONBAR_BACK and CustomBackColor (refer to the UCL.Colors.SelectThemeSet function).
Setting CustomBackColor.Enabled to true will use the color of CustomBackColor, otherwise, the color in CAPTIONBAR_BACK will be used.

**Comment by srd-software** [comment]
Thanks so much for the help. I figured the approach of using caption_back and customback was probably correct since it worked so well in my test app. 

Yes, I'll take a closer look at the colors unit. I used some of that with a SynEdit control to store its default colors and it worked well for app activate and deactivate along with the captionbar. I thought that might be something to explore further and you have confirmed that.

So I'll look at and experiment with the themecolorset stuff today. I'd like to leave this question open for now as I may have a few more questions when I get into the themecolorsets.

Thanks again!
